sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -1882,3 +1882,141 @@ case class ArrayRepeat(left: Expression, right: Expression)
+      val getValue = CodeGenerator.getValue(array, elementType, i)
+      val openHashSet = classOf[OpenHashSet[_]].getName
+      val classTag = s"scala.reflect.ClassTag$$.MODULE$$.Object()"
+      s"""
+         |int $distinctArrayLen = 0;
+         |$openHashSet $hs = new $openHashSet($classTag);
+         |for (int $i = 0; $i < $array.numElements(); $i++) {
+         |  if ($array.isNullAt($i)) {
+         |    int $j;
+         |    for ($j = 0; $j < $i; $j ++) {
+         |      if ($array.isNullAt($j))
+         |        break;
+         |    }
+         |    if ($i == $j) {
+         |      $distinctArrayLen = $distinctArrayLen + 1;
+         |    }
+         |  }
+         |  else {
+         |    if (!($hs.contains($getValue))) {
+         |      $hs.add($getValue);
+         |      $distinctArrayLen = $distinctArrayLen + 1;
+         |    }
+         |  }
+         |}
+         |${genCodeForResult(ctx, ev, array, distinctArrayLen)}
+      """.stripMargin
+    })
+  }
+
+  def genCodeForResult(
+      ctx: CodegenContext,
+      ev: ExprCode,
+      inputArray: String,
+      newArraySize: String): String = {
+    val distinctArr = ctx.freshName("distinctArray")
+    val hs = ctx.freshName("hs")
+    val openHashSet = classOf[OpenHashSet[_]].getName
+    val i = ctx.freshName("i")
+    val j = ctx.freshName("j")
+    val pos = ctx.freshName("pos")
+    val genericArrayData = classOf[GenericArrayData].getName
+    val getValue = CodeGenerator.getValue(inputArray, elementType, i)
+
+    if (!CodeGenerator.isPrimitiveType(elementType)) {
+      val arrayClass = classOf[GenericArrayData].getName
+      val classTag = s"scala.reflect.ClassTag$$.MODULE$$.Object()"
+      s"""
+         |Object[] $distinctArr = new Object[$newArraySize];
+         |int $pos = 0;
+         |$openHashSet $hs = new $openHashSet($classTag);
+         |for (int $i = 0; $i < $inputArray.numElements(); $i++) {
+         |  if ($inputArray.isNullAt($i)) {
+         |    int $j;
+         |    for ($j = 0; $j < $i; $j ++) {
+         |      if ($inputArray.isNullAt($j))
+         |        break;
+         |    }
+         |    if ($i == $j) {
+         |      $distinctArr[$pos] = null;
+         |      $pos = $pos + 1;
+         |    }
+         |  }
+         |  else {
+         |    if (!($hs.contains($getValue))) {
+         |      $hs.add($getValue);
+         |      $distinctArr[$pos] = $getValue;
+         |      $pos = $pos + 1;
+         |    }
+         |  }
+         |}
+         |${ev.value} = new $arrayClass($distinctArr);
+       """.stripMargin
+    } else {
+      val primitiveValueTypeName = CodeGenerator.primitiveTypeName(elementType)
+      val classTag = s"scala.reflect.ClassTag$$.MODULE$$.$primitiveValueTypeName()"
+      s"""
+         |${ctx.createUnsafeArray(distinctArr, newArraySize, elementType, s" $prettyName failed.")}
+         |int $pos = 0;
+         |$openHashSet $hs = new $openHashSet($classTag);
```

Comment:
Does it work for `boolean`, `float`, or `double`?

Reply:
Yes, It works for ```boolean```, ```float``` and ```double```. Here are the generated code: 

```org.apache.spark.util.collection.OpenHashSet project_hs_1 = new org.apache.spark.util.collection.OpenHashSet(scala.reflect.ClassTag$.MODULE$.Boolean());```

```org.apache.spark.util.collection.OpenHashSet project_hs_1 = new org.apache.spark.util.collection.OpenHashSet(scala.reflect.ClassTag$.MODULE$.Double());```

```org.apache.spark.util.collection.OpenHashSet project_hs_1 = new org.apache.spark.util.collection.OpenHashSet(scala.reflect.ClassTag$.MODULE$.Float());```

I also added UT for ```boolean```, ```float``` and ```double```.

Reply:
Sorry, it should work. But, it is not specialized for `boolean`, `float`, or `double`. nvm.
Thanks.